Added Feature Tag called parallelize

tmulkern opened the pull request:
Added a tag called parallelize which which is tag that works at a Feature level. It will generate code for the Feature to run explicitly in parallel for Munit3, Nunit3 and Xunit2.

- For **Nunit3** it will add `ParallelizableAttribute`, which as described here https://github.com/nunit/docs/wiki/Parallelizable-Attribute.

- For **Mbunit3** it will add `ParallelizableAttribute`, which as described here http://blog.bits-in-motion.com/2009/03/announcing-gallio-and-mbunit-v306.html?m=1

- For **Xunit2** it will added a `CollectionAttribute` with a unique name (GUID) as described here
https://xunit.github.io/docs/running-tests-in-parallel.html

This will allow for Mixing for Features that can be run in Parallel with ones that cannot, it will also add support for enabling MUnit3 generated tests to run in parallel because unlike Nunit3 use of ParallelScope, MUnit3 must explicitly set which TestFixtures can run in parallel

SabotageAndi commented:
Thanks for the PR. I think this is a good idea to have such a feature, but I am not sure about the name.
Perhaps we should but such special tags in a "namespace" so that if someone is already using this tag, has with an update new, unexpected behaviour.
I was thinking in that case about @specflow:parallelize. @gasparnagy any thoughts on this?

@tmulkern Could you fix the `parrallelize`- typos (double r) in the code before we have a depper look at it? Somehow it iritates me when I am reading the code. Thanks!

tmulkern commented:
@SabotageAndi Thanks for the feedback, I know what you mean about typos, will fix them up

tmulkern commented:
@SabotageAndi I also agree with the namespacing

tmulkern commented:
@SabotageAndi I have fixed the spelling mistake

SabotageAndi commented:
Thanks! Looks good for me.
@gasparnagy : Your 2 cents? Can we merge it?

tmulkern commented:
@SabotageAndi will I add the namespacing as you suggested?

tmulkern commented:
@gasparnagy bump

godrose commented:
Ohhh that's one magnificent feature. I had to split features to introduce
parallelization. When do you plan to distribute it via nuget?

On 24 Nov 2016 23:03, "Tadhg Mulkern" <notifications@github.com> wrote:

> Added a tag called parallelize which which is tag that works at a Feature
> level. It will generate code for the Feature to run explicitly in parrallel
> for Munit3, Nunit2 and Xunit2.
>
>    -
>
>    For *Nunit3* it will add ParallelizableAttribute, which as described
>    here https://github.com/nunit/docs/wiki/Parallelizable-Attribute.
>    -
>
>    For *Mbunit3* it will add ParallelizableAttribute, which as described
>    here http://blog.bits-in-motion.com/2009/03/announcing-gallio-
>    and-mbunit-v306.html?m=1
>    -
>
>    For *Xunit2* it will added a CollectionAttribute with a unique name
>    (GUID) as described here
>    https://xunit.github.io/docs/running-tests-in-parallel.html
>
> This will allow for Mixing for Features that can be run in Parallel with
> ones that cannot, it will also add support for enabling MUnit3 generated
> tests to run in parrallel because unlike Nunit3 use of ParallelScope,
> MUnit3 must explicitly set which TestFixtures can run in parrallel
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/techtalk/SpecFlow/pull/746
> Commit Summary
>
>    - Merge remote-tracking branch 'refs/remotes/techtalk/master' into
>    tag_parallelize
>    - Add ParrallelizeDecarator for NUnit3, MbUnit3 and XUnit2
>
> File Changes
>
>    - *M* TechTalk.SpecFlow.Generator/DefaultDependencyProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-0> (1)
>    - *M* TechTalk.SpecFlow.Generator/TechTalk.SpecFlow.Generator.csproj
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-1> (1)
>    - *A* TechTalk.SpecFlow.Generator/UnitTestConverter/
>    ParrallelizeDecorator.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-2> (41)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    IUnitTestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-3> (1)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    MbUnit3TestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-4> (7)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    MbUnitTestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-5> (5)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    MsTestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-6> (5)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    NUnit3TestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-7> (6)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    NUnitTestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-8> (5)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    XUnit2TestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-9> (6)
>    - *M* TechTalk.SpecFlow.Generator/UnitTestProvider/
>    XUnitTestGeneratorProvider.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-10> (7)
>    - *M* Tests/TechTalk.SpecFlow.GeneratorTests/
>    NUnit3GeneratorProviderTests.cs
>    <https://github.com/techtalk/SpecFlow/pull/746/files#diff-11> (15)
>
> Patch Links:
>
>    - https://github.com/techtalk/SpecFlow/pull/746.patch
>    - https://github.com/techtalk/SpecFlow/pull/746.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/techtalk/SpecFlow/pull/746>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHbkI_P4Obu-CIZe-rH2uosNkddZ227-ks5rBfuQgaJpZM4K79cX>
> .
>

tmulkern commented:
@godrose I am just waiting for the OK on this PR and get it merged. Hopefully it'll make the 2.2 release

gasparnagy commented:
@tmulkern @godrose @SabotageAndi I generally think that this is something that would be useful for many people, but I adding predefined English tag names is problematic. The problem is that while many people use SpecFlow with English gherkin files, there are a lot of projects that use these scenarios to have a better conversation with their stakeholders in a non-English project. And you should also think about languages that do not even use Latin letters, like Russian or Chinese.
Therefore our goal was to leave the choice for choosing the tag that is fitting to their communication model, and somehow let them map it to concrete functionality. (As you might have spotted out, there is already one that was predefined, the `@ignore` tag, but this is rather a bad example from this perspective.)

I am not saying that we should turn this idea down, but we should think about what other options we have that works better for non-English gherkin files. (Of course we could turn this into a plugin, so that it is not part of the core and people can either use it as it is or make their own version with their own tag name, but the problem we solve here is not plugin-like, as the parallelization is part of the core anyway.)

Also this way of attribution represents a kind of opt-in model, so you need to mark those features/scenarios that you want to run parallel, where in many cases an opt-out model works better.

Could you maybe describe the situation where you needed to separate the parallel and non-parallel running tests? Maybe seeing a concrete example might also help to see what would be the best way to address this problem.

tmulkern commented:
@gasparnagy Thanks for getting back to me

- I take your point on the different languages, but as you pointed out the ignore tag is already there, since parallelization is part of the core, I didn't think this would be an issue having this tag in English

- I can easily change this to be an opt-out model with a `@DoNotRunInParallel` tag instead? Then create an assembly level attribute that can turn on parallelization that can be used for all frameworks?

- I don't think the example will help, I think is is a simple enough to understand without an example. I am just trying to emulate in Gherkin which is already functionality of the frameworks mentioned.

tmulkern commented:
@gasparnagy 
Maybe even better would be an assembly level attribute that specifies the opt-out tag in the stake holders language? Something like `[assembly: SpecflowParallelize("DoNotRunInParallel")]`

That way it solves the language issue

tmulkern commented:
@gasparnagy Or if not an assembly attribute, then a conflict section element in the app.config?

SabotageAndi commented:
@gasparnagy I think making parallelization opt-out will break a lot of existing tests, when SpecFlow gets updated there.
First the users get propably a lot of errors because they were using ScenarioContext.Current which is then not more possible.
Second, the chance that the tests and the tested software are not written for parallel tests is high. This will result in conflicts.

I have the opinion, that how the scenarios are executed is a job of the test runner and not of SpecFlow. If it runs the Scenarios per default in parallel, thats fine and the user knows that. If  it runs one Scenario at a time, it is also fine.
AFAIK enabling parallel execution is opt-in on NUnit, so why should me make an opt-out of that. This is a surprise behavior difference for the user.

@tmulkern 
Assembly attributes will not work, because we need the information at generation time. Config- section would be fine for me. 
Addtionally we could provide standard values for the different languages.

tmulkern commented:
@SabotageAndi my idea was not to enable parallelization by default and then selectively opt-out the test cases, but instead have a configuration entry that would switch on the generation of parallelization code (for the frameworks mentioned above), with an opt-out tag that is specified in the config that will disable the generation of parallelization code for that 'tagged' feature. 

Example (open to suggestions for the attribute names):

` <generator 
      allowDebugGeneratedFiles="false" 
      allowRowTests="true"
      generateAsyncTests="false"
      generateParallelizationCode="true"
      doNotRunInParallelTag="DoNotRunInParallel"
 />
`

In regards to the standard values for the different languages I would not be in favor as it is much easier for the user to specify the tag in the config. That way there would not be the additional overhead of standard values for `n` languages and `n` charsets 


The only way that I know of enabling parallelization with out generating any additional code right now (correct me if I am wrong) is by using Nunit3 and adding the an assembly level attribute as stated in the wiki https://github.com/techtalk/SpecFlow/wiki/Parallel-Execution#notes-for-nunit-v3-support. 

To enable parallelization in XUnit2 requires adding the `CollectionAttribute` and for Mbunit3 requires adding the `ParallelizableAttribute`

TBH I originally started this PR to enable parallelization for Mbunit, but I decided to come up with a way to implement it for all the Frameworks that can support parallelization

@SabotageAndi @gasparnagy If you are ok with my suggestions (from this comment) I will make changes to PR to reflect the conversation.

gasparnagy commented:
@tmulkern @SabotageAndi thx for the feedback. 

@tmulkern I don't agree with you about the value of an example. Examples are the driving forces in BDD and help us to implement something that is really valuable for our users and not just features that we have to maintain. I have also done this mistake and I could list several features in SpecFlow that are actually not used by anyone, but causing troubles, like the BeforeScenarioBlock hook. But let's get back to our topic.

I can imagine the following scenario (this is an example):

- you have an existing project with many scenarios, execution time is too long
- you switch to xUnit 2 and everything starts to break
- you switch off parallel execution in xUnit, because you have realized that your project is not ready yet for parallel
- you work on making the project parallelizable, you fix a large part of it, so you could benefit from the faster execution, but there is a small portion that depends on an external dependency that you cannot test parallel
- you re-enable parallel in xUnit but you would like to exclude those scenarios that need that external dependency

At this point this example would lead to the opt-out model, but actually I would not even need a `@parallel` attribute here, because my scenarios that have been tagged with the `@myspecialexternaldep` are the ones that I would like to exclude (for now at least).

So if we start from this, we should rather implement this feature in a way, that you can specify a **list of tags** in the specflow configuration and if any of them is used in a scenario, we should disable the parallel execution.

Maybe from another example we should get to another solution, this is why I would be really interested in what was the example (in the level I was listing it), that have made you starting this PR. I would be also interested in yours, of course, @SabotageAndi.

tmulkern commented:
@gasparnagy I think I mistook your request for an example, I though you were looking for some examples in Gherkin, which would have a bit of a bother as the usage cases I have in Gherkin are subject to NDA etc. But the format you have described, I am happy to oblige, sorry about the confusion ;-)

As it happens the scenario you described is exactly the scenario I am having with MBunit. Right now it's an all or nothing approach with running in Features in parallel, so opting-out tags would be fine with me. The reason why I started with an opt-in model is that it was simpler to implement, but after your feedback I see that the opt-out model, with user specified tags is a much more useful model.

I'll get started on right the code for this, thanks.

tmulkern commented:
Ok, made the changes to the code, this is a example of what the config would look like

```
<generator allowDebugGeneratedFiles="false"
                           generateParallelCodeForFeatures="true">
                    <ignoreParallelTags>
                        <tag value="mySpecialTag1"/>
                        <tag value="mySpecialTag2"/>
                    </ignoreParallelTags>
</generator>
```

tmulkern commented:
@gasparnagy @SabotageAndi Does the updated code look good?

SabotageAndi commented:
@gasparnagy? From the code perspective it looks good.

gasparnagy commented:
@tmulkern @SabotageAndi I will give feedback tomorrow (today is very busy).

tmulkern commented:
@gasparnagy Thanks for the update, appreciate it

gasparnagy commented on TechTalk.SpecFlow.Generator/UnitTestConverter/DoNotParallelizeDecorator.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public void DecorateFrom(string tagName, TestClassGenerationContext generationContext)
+        {
+            generationContext.UnitTestGeneratorProvider.GenerateParallelCodeForFeature = false;
```
The unit test providers are singleton and should be stateless. I understand that you reset this flag for every class generation, but still this should be avoided.
You can use the TestClassGenerationContext for keeping this state (there is a CustomData property for such things, but you can also put your own flag into it). 
But seeing the code, I am thinking whether the decorator really helps here. As we need to modify all test porviders and the UnitTestFeatureGenerator as well, I wonder if it would be easier to put the decision logic (that is currently in the CanDecorateFrom method here) into a method in the UnitTestFeatureGenerator and just call it from the end of the SetupTestClass where you read the bool property. With that we could get rid of the bool property and the entire decorator class as well.
What do you think?

tmulkern commented on TechTalk.SpecFlow.Generator/Configuration/GeneratorConfigurationProvider.cs:
```diff
@@ -50,7 +50,7 @@ public IEnumerable<PluginDescriptor> GetPlugins(SpecFlowConfigurationHolder conf
             catch(Exception ex)
             {
-                throw new ConfigurationErrorsException("SpecFlow configuration error", ex);
+                throw new ConfigurationErrorsException(string.Format("SpecFlow configuration error :{0}", ex), ex);
```
Same answer...

tmulkern commented:
@gasparnagy As regards the config, I just keep it the same as all the other config entries regarding lists

https://github.com/techtalk/SpecFlow/blob/21b7aa7bd0664927e3b48ca92c9b6563f06e9e53/TechTalk.SpecFlow/PlatformSpecific/ConfigurationSectionHandler.cs#L159

https://github.com/techtalk/SpecFlow/blob/21b7aa7bd0664927e3b48ca92c9b6563f06e9e53/TechTalk.SpecFlow/PlatformSpecific/ConfigurationSectionHandler.cs#L193

gasparnagy commented:
@tmulkern on the config. Yes, I understood your goal, and I am also not 100% sure. But the dependencies and the step assemblies are elements that are more complex objects and might be rooting from independent sources. In case of the excluded tags, it is more like a simple list that is defined per project. In most of the cases -- as we have seen -- there are probably only one tag there. That's why I suggested the comma separated list. But I am also not sure. @SabotageAndi what do you think?

SabotageAndi commented:
I am not for the comma separated list. 
We have a nice structured XML (and hopefully soon JSON ;-) ) and then we hack something like that. With such separated lists, it is always a question what is the separator. Is it comma or semicolon?

But I would suggest to make it a little bit more resusable and name it TagCollection and TagElement. When we start introducing such config options, we will start to use it more often in the future.
Like make the "ignore" tag also customizeable in this way.

gasparnagy commented:
@SabotageAndi would you prefer a separate tag for the parallel options, or embed it to the generator. Ie the way how @tmulkern did it now, or something like 

```
<generator ...>
  <parallel enable="true">
    <excludeTags>
      <tag name="tag1" />
    </excludeTags>
  </parallel>
</generator>
```

tmulkern commented:
@gasparnagy Made some changes based on your suggestions, i'm keep all the logic in (the now renamed) ParallelizeDecorator, that way I don't have to mess around with keeping state in TestClassGenerationContext or the UnitTestFeatureGenerator (keeping it stateless as you suggested) and removing any logic within UnitTestFeatureGenerator (I find there is a lot of code there already). This way all the logic is in one class, leave the UnitTestFeatureGenerator's to generate the code.

@SabotageAndi I have renamed  IgnoreParallelTagCollection to TagCollection and
IgnoreParallelTagElement to TagElement for reuse as suggested

gasparnagy commented:
@tmulkern this is nice! One question: why do you have to register it both as `ITestClassTagDecorator` and `ITestClassDecorator`?

SabotageAndi commented:
@gasparnagy I am fine with the current xml structure.

gasparnagy commented:
@SabotageAndi what does 'current' means for you?

SabotageAndi commented:
@gasparnagy Sorry, to be so unspecific. :-) I meant the one that is now implemented in this PR.

tmulkern commented:
@gasparnagy @SabotageAndi I noticed a mistake in my code when you asked about  why do you have to register it both as `ITestClassTagDecorator` and `ITestClassDecorator`. I'll update and get back to you

tmulkern commented:
@gasparnagy @SabotageAndi Just added the fix just to summarize what I did : 
Changed ParallelizeDecorator to call UnitTestGeneratorProvider.SetTestClassParallelize in ITestClassDecorator.DecorateFrom, it was called in ITestClassTagDecorator.DecorateFrom but that would mean it would be called for each tag that is not in the ignoreParallelCodeGenerationTagslist. Now all the ITestClassTagDecorator implementation does is remove the ignoreParallelCodeGenerationTags from the overall tags

gasparnagy commented:
@tmulkern Thx! After this change the `ITestClassTagDecorator` interface should not be required, right?

gasparnagy commented:
@tmulkern @SabotageAndi Configuration. OK. Let's stick to the current structure. Could we still do some renaming: `generateParallelCodeForFeatures` to `markTestClassesParallelizable` (we don't generate parallel code here) and `ignoreParallelTags` to `skipParallelizableMarkerForTags` (we don't ignore the tags but rather the opposite; the word "ignore" might be misleading anyway)?

So the result would be

```
<generator allowDebugGeneratedFiles="false"
                   markTestClassesParallelizable="true">
                    <skipParallelizableMarkerForTags>
                        <tag value="mySpecialTag1"/>
                        <tag value="mySpecialTag2"/>
                    </skipParallelizableMarkerForTags>
</generator>
```

tmulkern commented:
@gasparnagy Will make those changes, the rest of the code ok with you?

tmulkern commented:
@gasparnagy On second reading of your comment, I think that renaming: `generateParallelCodeForFeatures` to `markTestFeaturesParallelizable` would make more sense, what do you think?

tmulkern commented:
@SabotageAndi Sorry, missed you earlier comment about `ITestClassTagDecorator`. I would like to keep it there as it would just cleanup up the tag's and remove the ones specified from the config from the overall list. I know I could do it using `ITestClassDecorator.DecorateFrom` method, but I would keep separation of concerns. 'ITestClassTagDecorator' to decorate the Test Class and `ITestClassTagDecorator` to cleanup the tags

gasparnagy commented:
@tmulkern "feature" vs "test class" - yes, I was also thinking on this. I am kind of OK with both. I have chosen the test class, because it better describes what is happening really. We could also say `enableFeatureLevelParallelizableConfiguration`... Finding the right name is hard... :)

tmulkern commented:
@gasparnagy I'll go with `markFeaturesParallelizable` and update the code

tmulkern commented:
@gasparnagy @SabotageAndi I have renamed `generateParallelCodeForFeatures ` to `markFeaturesParallelizable  `the config should look like this now

```
     <generator allowDebugGeneratedFiles=""false""
                           markFeaturesParallelizable=""true"">
                    <skipParallelizableMarkerForTags>
                        <tag value=""mySpecialTag1""/>
                        <tag value=""mySpecialTag2""/>
                    </skipParallelizableMarkerForTags>
     </generator>
```

tmulkern commented:
@gasparnagy Do I need to do anything more to this PR?

gasparnagy commented:
@tmulkern The only thing I have pending is the status of the `ITestClassTagDecorator` interface. From the code it looks like that it is not needed, so you can remove the interface implementation and the registration too. Or did I miss something?

tmulkern commented:
@gasparnagy You have missed something but it's not your fault, I addressed the comment to @SabotageAndi instead of you by mistake. Here is the comment

> @SabotageAndi Sorry, missed you earlier comment about ITestClassTagDecorator. I would like to keep it there as it would just cleanup up the tag's and remove the ones specified from the config from the overall list. I know I could do it using ITestClassDecorator.DecorateFrom method, but I would like to keep separation of concerns. ITestClassTagDecoratorto decorate the Test Class and ITestClassTagDecorator to cleanup the tags

tmulkern commented:
@gasparnagy @SabotageAndi Bump

SabotageAndi commented:
@gasparnagy Could we merge this now?

gasparnagy commented:
@tmulkern @SabotageAndi Back from holidays... 
@tmulkern: Regarding your comment on ITestClassTagDecorator / ITestClassDecorator: I don't really understand the comment. Currently the `DecorateFrom` of the `ITestClassTagDecorator` is empty. SpecFlow will call the `ITestClassTagDecorator.CanDecorateFrom`, but the result is only used to call a method that does not do anything (because it is empty). So this is all looks unnecessary. What do I miss?

tmulkern commented:
@gasparnagy @SabotageAndi Apologies for not getting back to this sooner, I was sick and had to catch up on work because of it.

@gasparnagy The reason I am implementing it as `ITestClassTagDecorator` is to remove the `ignoreParallelCodeGenerationTags` [L20](https://github.com/techtalk/SpecFlow/pull/746/files#diff-c3d419269ae2a10497ac2693b4df3d00R20) from the overall tags list for the Feature File using the `RemoveProcessedTags` Property [L38](https://github.com/techtalk/SpecFlow/pull/746/files#diff-c3d419269ae2a10497ac2693b4df3d00R38) if I didn't, the tags would be added as Test Categories. So you are correct, the `ITestClassTagDecorator.DecorateFrom` method is empty and does nothing, but I am using `ITestClassTagDecorator.CanDecorateFrom` to match the tag and `ITestClassTagDecorator.RemoveProcessedTags` to remove the tags from the overall list.

Maybe the question is, do I need to remove them from the overall tag list?

gasparnagy commented:
@tmulkern I see. If you remove the tags, they will not generate a category attribute (or similar depending on the unit test framework) for that particular tag. Since in our case the tags that we hook on (e.g. `@myspecialexternaldep`) are probably tags that are meaningful anyway also as a category, we should not "remove" them.

tmulkern commented:
@gasparnagy Ok then, I'll remove the `ITestClassTagDecorator` implementation. Thanks for getting back to me

tmulkern commented:
@gasparnagy All those changes are in

gasparnagy commented:
@tmulkern thx for the fast response. Somehow the line endings in the `DefaultDependencyProvider.cs` were changed in your one but last commit. It would be great to keep a proper history for this file, because many changes touch it. Could you somehow fix this and revert the line endings to what it was?

tmulkern commented:
@gasparnagy I didn't realise that, must be because I edited that file using GitHub's online tools. I must remember to do all my file changes on my computer from now on 😉.

I'll fix the file as soon as I can

tmulkern commented:
@gasparnagy Ok, changes in

gasparnagy commented:
@tmulkern Thx for the contribution! I will merge it now.

tmulkern commented:
Great stuff, thanks

tmulkern commented:
@godrose pull request is merged, just to give you a heads up

brettveenstra commented:
in the CI build (ci583), when I turn on parallelization of feature files, the NUnit parallelization values use the Feature file names instead of values like ParallelScope.None, or ParallelScope.Fixtures, etc.

SabotageAndi commented:
@brettveenstra Could you please open an issue with this problem and attach a small sample where we can see the failure?
Thanks!

liamharries commented:
Sorry if this is the wrong place to post this, please redirect me if needed.
I'm trying to exclude a test from parallel exection and having some difficulty.
I've included the generator as mentioned into my app.config and applied the tag to my scenario
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <configSections>
    <section name="specFlow" type="TechTalk.SpecFlow.Configuration.ConfigurationSectionHandler, TechTalk.SpecFlow"/>
  </configSections>
  <appSettings>
    ...
  </appSettings>
  <specFlow>
    <!-- For additional details on SpecFlow configuration options see http://go.specflow.org/doc-config -->
    <unitTestProvider name="SpecRun"/>
    <plugins>
      <add name="SpecRun"/>
    </plugins>
    <generator allowDebugGeneratedFiles="false"
               markFeaturesParallelizable="false">
      <skipParallelizableMarkerForTags>
        <tag value="Sequential"/>
        <tag value="sequential"/>
      </skipParallelizableMarkerForTags>
    </generator>
  </specFlow>
  <startup>
    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.2"/>
  </startup>
</configuration>
```
The test however still runs in parallel :(

Does anyone know how i can get my test to be excluded from parallel execution?

Below is my test

```
@Authentication
	Feature: Authentication
	So that only privileged users may access the system
	As the product owner
	I want the product to be authenticated
	
@AllStandard
Scenario Outline: Login
	Given I am at the <page1> page
	When I provide a <username> username and a <password> password
	Then I am at the <page2> page
	And unsuccessful login attempt pop-up is <visibility>

	Examples: 
	| description       | page1 | username | password | page2    | visibility |
	| valid credentials | login | true     | true     | Services | false      |
	| invalid username  | login | false    | true     | login    | true       |
	
@AllStandard
@Sequential
Scenario Outline: LoginSequential
	Given I am at the <page1> page
	When I provide a <username> username and a <password> password
	Then I am at the <page2> page
	And unsuccessful login attempt pop-up is <visibility>

	Examples: 
	| description       | page1 | username | password | page2    | visibility |
	| invalid password  | login | true     | false    | login    | true       |
```
And my default.srprofile
```
<?xml version="1.0" encoding="utf-8"?>
<TestProfile xmlns="http://www.specflow.org/schemas/plus/TestProfile/1.5">
  <Settings projectName="iPipelineSpecTests" projectId="{5f94e355-e6ef-4c08-9807-0b89016244a7}" />
  <Execution stopAfterFailures="0"
             testThreadCount="8"
             testSchedulingMode="Random"
             apartmentState="MTA"
             retryFor="Failing"
             retryCount="0"/>
  <TestAssemblyPaths>
    <TestAssemblyPath>iPipelineSpecTests.dll</TestAssemblyPath>
  </TestAssemblyPaths>

  <Targets>

    <Target name="IE_11_Windows10">
      <Filter>@AllStandard | @Desktop_All | @IE_All | @IE_11_Windows10</Filter>
      <DeploymentTransformationSteps>
        <RelocateConfigurationFile target="Thread{TestThreadId}config.config"/>
        <ConfigFileTransformation configFile="App.config" >
          <Transformation>
            <![CDATA[<?xml version="1.0" encoding="utf-8"?>
							<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
                              <appSettings>
                                <add key="browser" value="internet explorer" xdt:Locator="Match(key)" xdt:Transform="SetAttributes(value)" />
                                <add key="version" value="11" xdt:Locator="Match(key)" xdt:Transform="SetAttributes(value)" />
                                <add key="platform" value="Windows 10" xdt:Locator="Match(key)" xdt:Transform="SetAttributes(value)" />
                              </appSettings>
							</configuration>
						]]>
          </Transformation>
        </ConfigFileTransformation>
      </DeploymentTransformationSteps>
    </Target>

         ...

  </Targets>
</TestProfile>

```
